packages/firestore/test/integration/api/server_timestamp.test.ts

```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
             return Promise.resolve();
           })
         )
-        .then(() => waitForRemoteEvent());
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return estimated value', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsAreEstimates(snapshot))
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return previous value', () => {
+    let previousSnapshot: firestore.DocumentSnapshot;
+
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          previousSnapshot = snapshot;
+          verifyTimestampsAreResolved(snapshot);
+        })
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot =>
+          verifyTimestampsUsePreviousValue(snapshot, previousSnapshot)
+        )
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return previous value of different type', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          verifyTimestampsAreResolved(snapshot);
+        })
+        .then(() =>
+          docRef.update('a', firebase.firestore.FieldValue.serverTimestamp())
+        )
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          expect(snapshot.get('a', { serverTimestamps: 'none' })).to.equal(
+            null
+          );
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
+          expect(snapshot.get('a', { serverTimestamps: 'previous' })).to.equal(
+            42
+          );
+        })
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          expect(
+            snapshot.get('a', { serverTimestamps: 'none' })
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
```

Comment:
Again, I'd prefer not throw in extra expectations just because it's convenient.  Testing the behavior of 'estimate' and 'previous' on actual resolved timestamps (aka normal Dates) is probably good, but should go elsewhere (either in the existing 'can return {previous,estimate}' tests or in a new test.

Reply:
Removed